src/validation.cpp

```diff
@@ -1613,8 +1613,10 @@ static unsigned int GetBlockScriptFlags(const CBlockIndex* pindex, const Consens
 
     unsigned int flags = SCRIPT_VERIFY_NONE;
 
-    // Start enforcing P2SH (BIP16)
-    if (pindex->nHeight >= consensusparams.BIP16Height) {
+    // BIP16 didn't become active until Apr 1 2012
+    // However, only one historical block violated the P2SH rules, so for simplicity, always leave P2SH
+    // on except for the one violating block.
+    if (!(pindex->phashBlock != nullptr && (*(pindex->phashBlock) == consensusparams.BIP16Exception))) {
```

Comment:
In commit "Use P2SH consensus rules for all blocks"

Maybe check BIP16Exception.IsNull() to be a little more strict. Could also remove unnecessary nesting and negatives:

```c++
if (consensusparams.BIP16Exception.IsNull() ||
    pindex->phashBlock == nullptr ||
    *pindex->phashBlock != consensusparams.BIP16Exception)
```